magefile.go

```diff
@@ -114,11 +139,234 @@ func (Build) All() {
 	)
 }
 
+func (Build) BuildImages() {
+	mg.Deps(
+		mg.F(Build.imageBuild, "addon-operator-manager"),
+		mg.F(Build.imageBuild, "addon-operator-webhook"),
+		mg.F(Build.imageBuild, "api-mock"),
+		mg.F(Build.imageBuild, "addon-operator-index"), // also pushes bundle
+	)
+}
+
+func (Build) PushImages() {
+	mg.Deps(
+		mg.F(Build.imagePush, "addon-operator-manager"),
+		mg.F(Build.imagePush, "addon-operator-webhook"),
+		mg.F(Build.imagePush, "addon-operator-index"), // also pushes bundle
+	)
+}
+
 // Builds the docgen internal tool
 func (Build) Docgen() {
 	mg.Deps(mg.F(Build.cmd, "docgen", "", ""))
 }
 
+func (b Build) imageBuild(cmd string) error {
+	mg.Deps(
+		mg.F(Build.cmd, cmd, "linux", "amd64"),
+	)
+
+	// clean/prepare cache directory
+	imageCacheDir := path.Join(cacheDir, "image", cmd)
+	if err := os.RemoveAll(imageCacheDir); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("deleting image cache: %w", err)
+	}
+	if err := os.Remove(imageCacheDir + ".tar"); err != nil && !os.IsNotExist(err) {
+		return fmt.Errorf("deleting image cache: %w", err)
+	}
+	if err := os.MkdirAll(imageCacheDir, os.ModePerm); err != nil {
+		return fmt.Errorf("create image cache dir: %w", err)
+	}
+
+	switch cmd {
+	case "addon-operator-index":
+		return b.buildOLMIndexImage(imageCacheDir)
+
+	case "addon-operator-bundle":
+		return b.buildOLMBundleImage(imageCacheDir)
+
+	default:
+		return b.buildGenericImage(cmd, imageCacheDir)
+	}
+}
+
+// generic image build function, when the image just relies on
+// a static binary build from cmd/*
+func (Build) buildGenericImage(cmd, imageCacheDir string) error {
+	imageTag := imageURL(cmd)
+	for _, command := range [][]string{
+		// Copy files for build environment
+		{"cp", "-a",
+			"bin/linux_amd64/" + cmd,
+			imageCacheDir + "/" + cmd},
+		{"cp", "-a",
+			"config/docker/" + cmd + ".Dockerfile",
+			imageCacheDir + "/Dockerfile"},
+
+		// Build image!
+		{containerRuntime, "build", "-t", imageTag, imageCacheDir},
+		{containerRuntime, "image", "save",
+			"-o", imageCacheDir + ".tar", imageTag},
+	} {
+		if err := sh.Run(command[0], command[1:]...); err != nil {
+			return fmt.Errorf("running %q: %w", strings.Join(command, " "), err)
+		}
+	}
+	return nil
+}
+
+func (b Build) buildOLMIndexImage(imageCacheDir string) error {
+	mg.Deps(
+		Dependency.Opm,
+		mg.F(Build.imagePush, "bundle"),
+	)
+
+	if err := sh.Run("opm", "index", "add",
+		"--container-tool", containerRuntime,
+		"--bundles", imageURL("addon-operator-bundle"),
+		"--tag", imageURL("addon-operator-index")); err != nil {
+		return fmt.Errorf("runnign opm: %w", err)
+	}
+	return nil
+}
+
+func (b Build) buildOLMBundleImage(imageCacheDir string) error {
+	mg.Deps(
+		Build.TemplateAddonOperatorCSV,
+	)
+
+	imageTag := imageURL("addon-operator-bundle")
```

Comment:
Let's also clean/rm the cacheDir before running these?

`mkdir -p` silently does nothing if the directory already exists, unlike `mkdir` which atleast errs out.

Reply:
ohh, my bad! I must've skipped to see this 

https://github.com/openshift/addon-operator/blob/b4679763856941b8294495cc1023069dd88bb6e6/magefile.go#L169-L176